### PR TITLE
AP_DDS: Joystick Support 2

### PR DESCRIPTION
**Pull request by tizianofiorenzani**
I noticed that this Pull Request has been stuck for one year.
https://github.com/ArduPilot/ardupilot/pull/25334

Tagging @srmainwaring and @Ryanf55.

### CODE REVIEW

**Comment by Ryanf55** [comment]
Please add some tests (automated), or some instructions on how to test this in CLI. Once we've got those and we're happy, the commit history can be cleaned up. 

Nice job cleaning up all the previous actions. Looking good.

**Comment by Ryanf55** [comment]
We also need some idea on how this will interract if you are also sending normal RC input (SBUS from RC transmitter for example). If that gives complete authority to the companion computer, whether the two inputs fight, or something else.

**Comment by tizianofiorenzani** [comment]
Besides fixing the style, I realized that I need to expose the Joy interfaces through ROS2. I will add it to the same PR.

**Comment by tizianofiorenzani** [comment]
@Ryanf55 I am working on the tests. For the RC, we might check whether the Radio is connected before overriding its channels. This would prevent any conflict and prioritize RC at all time.

**Comment by Ryanf55** [comment]
Yep. It's worth reviewing this wiki and seeing whether we want to make DDS act in a similar way of "no pulses" and re-use the existing failover logic.
https://ardupilot.org/copter/docs/common-multiple-rx.html#failsafe-and-changeover
It's not clear how ArduPilot ranks the priority of two active receivers if they are both providing pules.

**Comment by Ryanf55** [comment]
Looks good, adding in the teleop joy instructions could be nice for this PR. I can dump it into the wiki. 

Can you clean up the history and fix your `git config user.name` to not be `=`?

**Comment by tizianofiorenzani** [comment]
@Ryanf55 of course, I forgot to update my laptop configuration.
Also, I have changed the code such that an input joystick value >= UINT16_MAX disables the channel override.

File: libraries/AP_DDS/AP_DDS_Client.cpp
Comment (Ryanf55) [inline libraries/AP_DDS/AP_DDS_Client.cpp:569]: Yea, seems right. Or, we could also just use NaN to mean invalid in ROS if it's documented. The UINT16_MAX is only because of mavlink using int's and not having NaN on that message, which shouldn't really affect this interface if there's better ways.

Comment (tizianofiorenzani) [reply libraries/AP_DDS/AP_DDS_Client.cpp]: I replaced it with a Nan and tested on the SITL.

File: libraries/AP_DDS/AP_DDS_Client.cpp
Comment (tizianofiorenzani) [inline libraries/AP_DDS/AP_DDS_Client.cpp:567]:
```suggestion
                // Ignore channels if NaN. 
```

File: Tools/GIT_Test/GIT_Success.txt
Comment (tizianofiorenzani) [inline Tools/GIT_Test/GIT_Success.txt]:
```suggestion
Tsuyoshi Arakawa
Naveen Kumar Kilaparthi
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp
Comment (tizianofiorenzani) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:51]:
```suggestion
    const float tmp = KELVIN_TO_C(_crude_value * 0.02);
```

File: Tools/GIT_Test/GIT_Success.txt
Comment (tizianofiorenzani) [inline Tools/GIT_Test/GIT_Success.txt:203]:
```suggestion
Tsuyoshi Arakawa
Naveen Kumar Kilaparthi
```

File: libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp
Comment (tizianofiorenzani) [inline libraries/AP_TemperatureSensor/AP_TemperatureSensor_MLX90614.cpp:51]:
```suggestion
    const float tmp = KELVIN_TO_C(_crude_value * 0.02);
```

File: libraries/AP_DDS/README.md
Comment (Ryanf55) [inline libraries/AP_DDS/README.md:257]: Add to `published topics` section on line 185 too.

File: libraries/AP_DDS/README.md
Comment (Ryanf55) [inline libraries/AP_DDS/README.md:256]:
```suggestion
## Commanding the vehicle using ROS 2 Topics
```

File: libraries/AP_DDS/README.md
Comment (Ryanf55) [inline libraries/AP_DDS/README.md:261]: Should add a note on how priority is determined. We also have the potential to let RC have priority with a switch it looks like.
https://github.com/ArduPilot/ardupilot/issues/7045#issuecomment-841997340

**Comment by Ryanf55** [comment]
I force pushed with clean history. Can you share which email you want to attribute the work to?

**Review by Ryanf55 — COMMENTED** [review]
Functionally seems to work fine. It clamps to -1/1 through the linear interpolate nicely. I think it's OK to silently clamp. 
image

I think it might be nicer in the readme to give a command for a right turn and climb in FBWB.

**Review by Ryanf55 — APPROVED** [review]

File: libraries/AP_DDS/README.md
Comment (tizianofiorenzani) [reply libraries/AP_DDS/README.md:259]: that is already mentioned in the Subscribed topics

**Comment by tizianofiorenzani** [comment]
@Ryanf55 The test is pushed and working now, run it with:
`colcon test --packages-select ardupilot_dds_tests --event-handlers=console_cohesion+ --pytest-args -k test_joy_msg_received`.

I also updated the Readme, I think we are good to go once CI has passed.

File: libraries/AP_DDS/README.md
Comment (tizianofiorenzani) [reply libraries/AP_DDS/README.md:265]: Added a comment to explain that each channel defaults to RC after 1s of inactivity.

**Comment by tizianofiorenzani** [comment]
@Ryanf55  I just noticed that when I passed the astyle on the Readme, it really messed it up. I will need to restore it and push it again.

**Comment by Ryanf55** [comment]
> * ./Tools/scripts/run_astyle.py

How did you run `astyle`? This doesn't run on the README. astyle is a C++ formatter, not a markdown formatter.

**Comment by tizianofiorenzani** [comment]
As a simple test, you can 

- run the ros2_sitl, quadcopter is fine, then takeoff. 
- On a new terminal, start sending joystick axes with `ros2 topic pub /ap/joy sensor_msgs/msg/Joy "{axes: [0.0, 0.0, 0.0, 0.0]}"`
- Set the vehicle in LOITER mode
- change the published joystick to `ros2 topic pub /ap/joy sensor_msgs/msg/Joy "{axes: [0.5, 0.0, 0.0, 0.0]}"` and you will see the vehicle moving sideways.

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py:68]: Remove commented code

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py
Comment (Ryanf55) [inline Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py:95]: Remove commented code.

**Review by Ryanf55 — APPROVED** [review]
Nice, looks good! Remove dead code and squash all your commits to one `AP_DDS: ...` commit.

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py
Comment (tizianofiorenzani) [reply Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py]: done, thanks

File: Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py
Comment (tizianofiorenzani) [reply Tools/ros2/ardupilot_dds_tests/test/ardupilot_dds_tests/test_joy_msg_received.py]: done, thanks

**Comment by tizianofiorenzani** [comment]
sitl-plane test fails. I don't know how to test it locally, here it's what is says:

2024-09-21T21:16:51.5884422Z >>>> FAILED STEP: test.Plane at Sat Sep 21 21:16:51 2024
2024-09-21T21:16:51.5885040Z Failure Summary:
2024-09-21T21:16:51.5885407Z   test.Plane:
2024-09-21T21:16:51.5887138Z     DeadreckoningNoAirSpeed (Test deadreckoning support with no airspeed sensor) (Failed to get EKF.flags=16) (see /tmp/buildlogs/ArduPlane-DeadreckoningNoAirSpeed.txt) (duration 10.50739574432373 s)

**Comment by Ryanf55** [comment]
> sitl-plane test fails. I don't know how to test it locally, here it's what is says:
> 
> 2024-09-21T21:16:51.5884422Z >>>> FAILED STEP: test.Plane at Sat Sep 21 21:16:51 2024 2024-09-21T21:16:51.5885040Z Failure Summary: 2024-09-21T21:16:51.5885407Z test.Plane: 2024-09-21T21:16:51.5887138Z DeadreckoningNoAirSpeed (Test deadreckoning support with no airspeed sensor) (Failed to get EKF.flags=16) (see /tmp/buildlogs/ArduPlane-DeadreckoningNoAirSpeed.txt) (duration 10.50739574432373 s)

Some tests are flaky. Don't worry about those passing till you have the commit history clean, and I'll rerun the flaky unrelated tests.

**Comment by tizianofiorenzani** [comment]
@Ryanf55 I rebased and cleaned up the history, leaving one single commit after your previous approval.

**Comment by Ryanf55** [comment]
I tagged for dev call for review, but will be late. Feel free to provide any feedback blocking merge.

File: libraries/AP_DDS/AP_DDS_Client.cpp
Comment (tridge) [inline libraries/AP_DDS/AP_DDS_Client.cpp:558]: can't send a text msg per joy msg, would flood the link

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_DDS/AP_DDS_Client.cpp
Comment (Ryanf55) [reply libraries/AP_DDS/AP_DDS_Client.cpp:558]: For context, ardupilot sends all message to the GCS by default. It's not like ROS where it prevents publish.

Comment (tizianofiorenzani) [reply libraries/AP_DDS/AP_DDS_Client.cpp]: done

Comment (tizianofiorenzani) [reply libraries/AP_DDS/AP_DDS_Client.cpp]: I removed all the GCS communications.

**Pull request merged**
